Source/DataModels/Components/UI/ComponentText2D.cpp

```diff
@@ -0,0 +1,267 @@
+	{
+		currentFont = resourceFont;
+	}
+#endif
+	text = meta["text"];
+	color.x = static_cast<float>(meta["color_x"]);
+	color.y = static_cast<float>(meta["color_y"]);
+	color.z = static_cast<float>(meta["color_z"]);
+}
+
+std::vector<std::shared_ptr<ResourceFont>> ComponentText2D::getFonts()
+{
+	std::vector<std::string> paths = App->GetModule<ModuleFileSystem>()->ListFiles("Assets/Fonts");
+	std::vector<std::shared_ptr<ResourceFont>> fonts;
+	for (std::string path : paths)
+	{
+		std::size_t found = path.find(".meta");
+		if (found != std::string::npos)
+		{
+			path.erase(path.begin() + found, path.end());
+			fonts.push_back(App->GetModule<ModuleResources>()->RequestResource<ResourceFont>(path));
+		}
+	}
+	return fonts;
+}
+
+void ComponentText2D::CreateVAO()
+{
+	glGenVertexArrays(1, &VAO);
+	glGenBuffers(1, &VBO);
+	glBindVertexArray(VAO);
+	glBindBuffer(GL_ARRAY_BUFFER, VBO);
+	glBufferData(GL_ARRAY_BUFFER, sizeof(float) * 6 * 4, NULL, GL_DYNAMIC_DRAW);
+	glEnableVertexAttribArray(0);
+	glVertexAttribPointer(0, 4, GL_FLOAT, GL_FALSE, 4 * sizeof(float), 0);
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+}
+
+void ComponentText2D::DeleteVAOandVBO()
+{
+	glDeleteVertexArrays(1, &VAO);
+	glDeleteBuffers(1, &VBO);
+}
+
+void ComponentText2D::RenderText(std::string text,
+								 float x,
+								 float y,
+								 float scale,
+								 float3 color,
+								 std::map<char, Character> characters) const
+{	
+	programShared->BindUniformInt("text", 0);
+	programShared->BindUniformFloat3("textColor", color);	
+	
+	glActiveTexture(GL_TEXTURE0);
+	glBindVertexArray(VAO);
+	/*ENGINE_LOG("VAO: %i", VAO)*/
+	// iterate through all characters
+	std::string::const_iterator c;
+	for (c = text.begin(); c != text.end(); c++)
+	{
+		Character ch = characters[*c];
+
+		float xpos = x + ch.Bearing.x * scale;
+		float ypos = y - (ch.Size.y - ch.Bearing.y) * scale;
+		
+		float w = ch.Size.x * scale;
+		float h = ch.Size.y * scale;		
+		// update VBO for each character
+		float vertices[6][4] = { { xpos / 48.0, (ypos + h) / 48.0, 1.0f, 1.0f },
+								 { xpos / 48.0, ypos / 48.0, 1.0f, 0.0f },
+								 { (xpos + w) / 48.0, ypos / 48.0, 0.0f, 0.0f },
+
+			{ xpos / 48.0, (ypos + h) / 48.0, 1.0f, 1.0f },
+								 { (xpos + w) / 48.0, ypos / 48.0, 0.0f, 0.0f },
+								 { (xpos + w) / 48.0, (ypos + h) / 48.0, 0.0f, 1.0f }
+		};
+		// render glyph texture over quad
+		glBindTexture(GL_TEXTURE_2D, ch.TextureID);
+		GLenum err;
+		while ((err = glGetError()) != GL_NO_ERROR)
+		{
+			std::string error;
+			switch (err)
+			{
+				case GL_INVALID_OPERATION:
+					error = "INVALID_OPERATION";
+					break;
+				case GL_INVALID_ENUM:
+					error = "INVALID_ENUM";
+					break;
+				case GL_INVALID_VALUE:
+					error = "INVALID_VALUE";
+					break;
+				case GL_OUT_OF_MEMORY:
+					error = "OUT_OF_MEMORY";
+					break;
+				case GL_INVALID_FRAMEBUFFER_OPERATION:
+					error = "INVALID_FRAMEBUFFER_OPERATION";
+					break;
+			}
+			LOG_INFO(error);
+		}
```

Comment:
1. Why use a `while` loop? You're never changing the value of `err` as far as I can tell
2. If it's an error, why log it as info?
```suggestion
		GLenum err = glGetError();
		if (err != GL_NO_ERROR)
		{
			std::string error;
			switch (err)
			{
				case GL_INVALID_OPERATION:
					error = "INVALID OPERATION";
					break;
				case GL_INVALID_ENUM:
					error = "INVALID ENUM";
					break;
				case GL_INVALID_VALUE:
					error = "INVALID VALUE";
					break;
				case GL_OUT_OF_MEMORY:
					error = "OUT OF MEMORY";
					break;
				case GL_INVALID_FRAMEBUFFER_OPERATION:
					error = "INVALID FRAMEBUFFER OPERATION";
					break;
				default:
					error = "UNKNOWN ERROR";
					break;
			}
			LOG_ERROR("OpenGL error while binding the texture in ComponentText2D owned by {}; error: {}", GetOwner(), error);
		}
```